baseapp/manager.go

```diff
@@ -2,113 +2,164 @@ package baseapp
 	"github.com/berachain/offchain-sdk/worker"
 )
 
-type JobManager struct {
-	// logger is the logger for the baseapp
-	logger log.Logger
+const (
+	producerName           = "job-producer"
+	producerPromName       = "job_producer"
+	producerResizeStrategy = "eager"
 
+	executorName     = "job-executor"
+	executorPromName = "job_executor"
+)
+
+// Manager handles the job and worker lifecycle.
+type Manager struct {
 	// list of jobs
-	jobs []job.Basic
+	jobRegistry *job.Registry
+
+	// Job producers are a pool of workers that produce jobs. These workers
+	// run in the background and produce jobs that are then consumed by the
+	// job executors.
+	producerCfg  *worker.PoolConfig
+	jobProducers *worker.Pool
 
-	// worker pool
-	executionPool worker.Pool
+	// Job executors are a pool of workers that execute jobs. These workers
+	// are fed jobs by the job producers.
+	executorCfg  *worker.PoolConfig
+	jobExecutors *worker.Pool
 }
 
-// New creates a new baseapp.
-func NewJobManager(
-	name string,
-	logger log.Logger,
+// NewManager creates a new manager.
+func NewManager(
 	jobs []job.Basic,
-) *JobManager {
-	return &JobManager{
-		logger: log.NewBlankLogger(os.Stdout),
-		jobs:   jobs,
-		executionPool: worker.NewPool(
-			name+"-execution",
-			4, //nolint:gomnd // hardcode 4 workers for now
-			logger,
-		),
+) *Manager {
+	m := &Manager{
+		jobRegistry: job.NewRegistry(),
 	}
-}
 
-// Start.
-//
-//nolint:gocognit // todo: fix.
-func (jm *JobManager) Start(ctx context.Context) {
-	for _, j := range jm.jobs {
-		if err := j.Start(ctx); err != nil {
+	for _, j := range jobs {
+		if err := m.jobRegistry.Register(j); err != nil {
 			panic(err)
 		}
+	}
 
-		//nolint:nestif // todo fix.
-		if condJob, ok := j.(job.Conditional); ok {
-			go func() {
-				for {
-					time.Sleep(50 * time.Millisecond) //nolint:gomnd // fix.
-					if condJob.Condition(ctx) {
-						jm.executionPool.AddTask(job.NewExecutor(ctx, condJob, nil))
-						return
+	jobCount := uint16(m.jobRegistry.Count())
+	m.producerCfg = &worker.PoolConfig{
+		Name:             producerName,
+		PrometheusPrefix: producerPromName,
+		MinWorkers:       jobCount,
+		MaxWorkers:       jobCount + 1,
+		ResizingStrategy: producerResizeStrategy,
+		MaxQueuedJobs:    jobCount,
+	}
+
+	// TODO: read from config.
+	m.executorCfg = worker.DefaultPoolConfig()
+	m.executorCfg.Name = executorName
+	m.executorCfg.PrometheusPrefix = executorPromName
+
+	return m
+}
+
+// Logger returns the logger for the baseapp.
+func (jm *Manager) Logger(ctx context.Context) log.Logger {
+	return sdk.UnwrapSdkContext(ctx).Logger().With("namespace", "job-manager")
+}
+
+// Start calls `Setup` on the jobs in the registry as well as spins up
+// the worker pools.
+func (jm *Manager) Start(ctx context.Context) {
+	// We pass in the context in order to handle cancelling the workers.
+	jm.jobExecutors = worker.NewPool(ctx, jm.executorCfg)
+	jm.jobProducers = worker.NewPool(ctx, jm.producerCfg)
```

Comment:
dont need context here

Reply:
have to pass to the underlying pool (for now until we handle new contexts and stuff)